### ISSUE TITLE
findtext returns empty string on integer zero value

### ISSUE THREAD

**Issue by bitsgalore** [python/cpython#91447]
ElementTree's "findtext" function returns an empty string value if the element's text field contains an integer with value 0. Below example illustrates the issue:

```python
import xml.etree.ElementTree as ET

# Create root element
test = ET.Element("test")

# Compute sum and difference on two numbers
firstNumber = 5000
secondNumber = 5000

sum = firstNumber + secondNumber
difference = firstNumber - secondNumber

# Create subelements for sum and difference values
el = ET.SubElement(test, "sum")
el.text = sum

el = ET.SubElement(test, "diff")
el.text = difference

# Print data types
print("type(sum): " + str(type(sum)))
print("type(diff): " + str(type(difference)))
print("type(sum subelt): " + str(type(test.findtext('./sum'))))
print("type(diff subelt): " + str(type(test.findtext('./diff'))))

# Print values of sub elements
print("sum subelt: " + str(test.findtext('./sum')))
print("diff subelt: " + str(test.findtext('./diff')))
```
This gives me the following output:

```
type(sum): <class 'int'>
type(diff): <class 'int'>
type(sum subelt): <class 'int'>
type(diff subelt): <class 'str'>
sum subelt: 10000
diff subelt: 
```
Note how the data type of the "diff" sub-element is "string", even though the source data is an integer.

I'm using Python 3.8.10 on Linux Mint 20.1 Ulyssa (based on Ubuntu Focal Fossa 20.04).

**Comment by eugenetriguba** [comment]
I anticipate this would happen for any "falsey" text attribute value. The `findtext` function is returning `elem.text or ""` and 0 is leading it to return the empty string. 

For instance:

```
>>> import xml.etree.ElementTree as ET
>>> test = ET.Element("test")
>>> el = ET.SubElement(test, "diff")
>>> el.text = False
>>> print(test.findtext('./diff'))

>>> 
```

You can stringify the value for the text attribute as a workaround.
```
>>> el.text = str(0)
>>> print(test.findtext('./diff'))
0
```

But I'm not familiar with why the current behavior is the way it is. From the API docs: "Note that if the matching element has no text content an empty string is returned." 

What is considered "empty"? Any reason not to do something like the following?

```
if elem.text is None:
    return ""

return elem.text
```

**Comment by bitsgalore** [comment]
@eugenetriguba Good call, hadn't tried the Boolean False myself! I ran into this problem while writing some [unit tests](https://github.com/KBNLresearch/isolyzer/blob/415b0bb585af4321f276199a9addb52206d152f9/tests/unit/test_testfiles.py#L82) that take a large element tree with low-level properties of ISO image files as input. Since these values are used internally by the software for various calculations, stringifying them isn't really an option, as that would make things unnecessarily complex.

My own interpretation of "no text content" would also be a "None" match. After posting this, I realized I'd come across this very same issue some ten years ago in another project (for reasons I can't remember I never got round to reporting it back then).  At the time we created a replacement for the *findText* function (link [here](https://github.com/openpreserve/jpylyzer/blob/94c11a5ef7033346dce0e4b06c0d4b5549438995/jpylyzer/etpatch.py#L56)), which is along the same lines as your suggestion.

**Comment by JelleZijlstra** [comment]
Is it even meaningful for elements to have an int or boolean as their `.text` attribute? I'd expect text to be either a string or None.

**Comment by eugenetriguba** [comment]
@JelleZijlstra I thought about that as well but felt like this was the less controversial change. 

It fixes what certainly seems like unintended and incorrect behavior. I assumed we could revisit the text attribute returning different types in a different issue since that is what the behavior seems to have been anyway. What are your thoughts there?

**Comment by eugenetriguba** [comment]
@JelleZijlstra Could you explain that? I'm not quite understanding your concern.

Currently falsey values (like the integer 0) return "" rather than the actual `.text` attribute that was set. That MR aims to resolve it so the falsey value is actually returned from `findtext` rather than giving an empty string for any falsey value. It doesn't change non-string values to be strings.

**Comment by JelleZijlstra** [comment]
> @JelleZijlstra Could you explain that? I'm not quite understanding your concern.
> 
> Currently falsey values (like the integer 0) return "" rather than the actual `.text` attribute that was set. That MR aims to resolve it so the falsey value is actually returned from `findtext` rather than giving an empty string for any falsey value. It doesn't change non-string values to be strings.

It changes user-visible behavior that users could be relying on. Sure, it's not very intuitive behavior and we wouldn't design it like this from scratch, but it's been like this for a long time, and there may well be code out there relying on this behavior.

**Comment by eugenetriguba** [comment]
@JelleZijlstra Is there someone or somewhere I could reach out to who could make a decision on the best way to move forward here? I tried sending an email to the core mentorship mailing list a bit over 2 weeks ago, but I got an automated email back that it is pending approval to be sent.

**Comment by JelleZijlstra** [comment]
Maybe start a discussion at discuss.python.org?

**Comment by gvanrossum** [comment]
The original test case here seems pretty awkward. Is this a proper demonstration of the issue?
```
~/cpython$ python3
Python 3.11.0b4 (v3.11.0b4:5a7e1e0a92, Jul 11 2022, 12:47:14) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import xml.etree.ElementTree as ET
>>> test = ET.Element("test")
>>> el = ET.SubElement(test, "sum")
>>> el.text = 42
>>> el = ET.SubElement(test, "diff")
>>> el.text = 0
>>> test.findtext('./sum')
42
>>> test.findtext('./diff')
''
>>> ^D
~/cpython$ 
```
IIUC you'd like the final output to be the integer `0`, not an empty string object. Right? Or is there a reason the example is so convoluted?

**Comment by bitsgalore** [comment]
@gvanrossum Yes, your example demonstrates the issue perfectly. Looking back at my original example, I agree it's a bit overly convoluted.

Also, as @eugenetriguba pointed out above, the same thing happens for other "falsey" text attributes, e.g.:

```
>>> el = ET.SubElement(test, "false")
>>> el.text = False
>>> test.findtext('./false')
''
```

**Comment by gvanrossum** [comment]
Honestly I think I should just merge your PR GH-91486 and be done with it. It will not land in 3.11 though, I don't feel this is a release blocker since the buggy behavior has been around for such a long time. We might fix it in 3.11.1 but that's up to the release manager.
